### PR TITLE
New application logging (zerolog) & error handling

### LINKED ISSUE #70: Drop logger, logLevel and use coraza's
Right now this connector spins up a new zap logger on bootstrap. Spawning a new zap logger for a connector is overkill, more so when the transactions carry a logger that are controlled by the directives log level (so it is consistent) and it is efficient when it comes to allocation. Removing a dependency is usually a good practice.

Lines https://github.com/corazawaf/coraza-spoa/blob/main/internal/spoa.go#L175-L193 should be removed. In case spoe has a logger we could plug that into WAF setup (e.g. https://github.com/corazawaf/coraza-caddy/blob/main/logger.go), otherwise we should follow and use the default WAF logger.

### PR DESCRIPTION
**Pull request by zc-devs**
Closes #70 

- Using zerolog for application logging
- Introduced application error responses: variables `err_code` and `err_msg` - SPOA errors
- `error` used for SPOE/SPOP errors

### CODE REVIEW

File: log/log.go [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+
+func Fatal() *zerolog.Event {
+	return Logger.Fatal()
+}
+func Panic() *zerolog.Event {
+	return Logger.Panic()
+}
+
+func resolveLogPath(path string) (io.Writer, error) {
+	if path == "" {
```
Comment (zc-devs) [inline log/log.go:93]: switch / case

File: internal/spoa.go [hunk 2]
```diff
@@ -291,7 +234,7 @@ func (s *SPOA) processRequest(spoeMsg *spoe.Message) ([]spoe.Action, error) {
 
 	tx = app.waf.NewTransactionWithID(req.id)
 	if tx.IsRuleEngineOff() {
-		app.logger.Warn("Rule engine is Off, Coraza is not going to process any rule")
+		log.Warn().Msg("Rule engine is Off, Coraza is not going to process any rule")
 		return s.message(miss), nil
 	}
```
Comment (zc-devs) [inline internal/spoa.go:240]: Log in application log too before throwing to criteo

Comment (zc-devs) [reply internal/spoa.go:240]: Find where `txn.coraza.error` is set.

Make constant error message like allowMessage. In case of error return that message here instead of criteo (if criteo is responsible for error messages now). Let criteo send error messages in case of his fail.

File: config.yaml.default [hunk 3]
```diff
@@ -1,6 +1,12 @@
-
 # The SPOA server bind address
 bind: 0.0.0.0:9000
+log:
+  # The log level configuration, one of: debug/info/warn/error/panic/fatal
+  level: trace
+  # The log file path, /dev/stderr by default
+  log_file: /dev/stdout
```
Comment (zc-devs) [inline config.yaml.default:7]: Just `file`

File: config.yaml.default [hunk 4]
```diff
@@ -1,6 +1,12 @@
...
+  # Log WAF errors as SPOA, default is false
+  log_waf: true
```
Comment (zc-devs) [inline config.yaml.default:9]: Maybe shorten to `waf`?

File: internal/spoa.go (hunk 2 above)
Comment (zc-devs) [reply internal/spoa.go:240]: https://github.com/corazawaf/coraza-spoa/blob/dd5eb86d1e9abbdd5fe568249f36a6d85257eba7/docker/haproxy/coraza.cfg#L5-L6

https://github.com/haproxy/haproxy/blob/49ddd87d41b62797d50d7c59e843d0c0b0236eb5/doc/SPOE.txt#L300

**Comment by zc-devs** [comment]
[log.txt](https://github.com/corazawaf/coraza-spoa/files/12061365/log.txt)

File: config/config.go [hunk 5]
```diff
@@ -32,6 +37,14 @@ type Application struct {
 	TransactionActiveLimit     int      `yaml:"transaction_active_limit"`
 }
 
+// Log is used to manage the SPOA logging.
+type Log struct {
+	Level     string `yaml:"level"`
```
Comment (jcchavezs) [inline config/config.go:42]: I don't think we should provide two ways to do login in a connector. Can we stick to tx's logger?

File: internal/spoa.go [hunk 6]
```diff
@@ -128,84 +154,34 @@ func (s *SPOA) readHeaders(headers string) (http.Header, error) {
...
 // New Create a new SPOA instance.
 func New(conf *config.Config) (*SPOA, error) {
 	apps := make(map[string]*application)
 	for name, cfg := range conf.Applications {
-		pe := zap.NewProductionEncoderConfig()
-
-		fileEncoder := zapcore.NewJSONEncoder(pe)
-
-		pe.EncodeTime = zapcore.ISO8601TimeEncoder
+		wafConf := coraza.NewWAFConfig().
```
Comment (jcchavezs) [inline internal/spoa.go:161]: I'd prefer we get the platform logging and plug into an implementation coraza can use. For example we use the zap logger created by caddy and plug into coraza https://github.com/corazawaf/coraza-caddy/blob/main/coraza.go#L54

**Comment by jcchavezs** [comment]
Thanks for raising this PR. I think in general I'd prefer we use the logger provided by the platform (in this case spoe) and inject it into coraza config to use a single/consistent logger. This will also prevent us from deviating on the log destination among connector and coraza itself. See for example what we do in caddy https://github.com/corazawaf/coraza-caddy/blob/main/coraza.go#L54

File: config/config.go (hunk 5 above)
Comment (zc-devs) [reply config/config.go:42]: No, we need application (`coraza-spoa`) logging. We need to log all initialization stuff as well as request/response before transaction is created.

File: internal/spoa.go (hunk 6 above)
Comment (zc-devs) [reply internal/spoa.go:161]: TBH, I had tried it before pushed the PR. It doesn't suit well.

**Comment by zc-devs** [comment]
1. As I said before, we need application logger to log stuff not related to request/response processing (transaction, coraza). Related stuff I log via transaction logger (coraza).
2. It good that Coraza can use external logger, but we can't reach `single/consistent logger` because of other libraries. At least because of `criteo/haproxy-spoe-go` now. Moreover, I think we should not.
3. Suppose, we plug `coraza-spoa` logger into `coraza`.
Who (which part) is responsible for configuration?
If it's `coraza-spoa`, then [SecDebugLog and SecDebugLogLevel ](https://github.com/corazawaf/coraza/blob/0dabae31d831f4f6d0f1bb57a1c5adc5cd4f2eb5/coraza.conf-recommended#L211-L212) would be ignored?
If it's `coraza`, then why `coraza-spoa` 's log is configured by other application/library?
As I understand, we plug in only debug logger (which contains errors, I guess). What's about audit log (`SecAuditLog`)?
4. As I've tried it, I want to distinguish `coraza-spoa` logs from underlying libraries logs. When I tried, `coraza-spoa` debug logs were 'lost' in a lot of `coraza`'s logs.
Ideally, it should be like [this](https://docs.spring.io/spring-boot/docs/2.1.13.RELEASE/reference/html/boot-features-logging.html#boot-features-custom-log-levels). But we have what we have.... Maybe later I will implement loggers by package/'class'.
5. There are two types of subsystems, I think:
* One is internal which is configured by application and _some_ of subsystem configuration may be exposed via application config itself. Configuration properties may differ from subsystem ones. Example: `coraza-spoa` - `haproxy-spoe-go`.
* Another is one, which is configured by its own config, which we (application) just supply to that subsystem. Example: `coraza-spoa` - `coraza` .

To sum up, `coraza-spoa` / `coraza` / `haproxy-spoe-go` is the different subsystems. `coraza` has its own configuration and should be configured through it.

**Comment by sts** [comment]
In the current version a configuration error will produce duplicated logs:

```
{"level":"error","error":"invalid WAF config from string: failed to parse string: failed to compile the directive \"secrule\": there is a another rule with id 10102","app":"test","time":"2023-07-29T14:42:14Z","message":"Unable to create WAF instance"}
{"level":"fatal","error":"invalid WAF config from string: failed to parse string: failed to compile the directive \"secrule\": there is a another rule with id 10102","time":"2023-07-29T14:42:14Z","message":"Can't initialize SPOA"}
```

How about we just log out the first line and not `return` in [internal/spoa](https://github.com/corazawaf/coraza-spoa/blob/main/internal/spoa.go#L188) but rather `continue`?

File: config/config.go [hunk 7]
```diff
@@ -54,11 +67,20 @@ func InitConfig(file string) error {
...
-		if app.LogLevel == "" {
-			app.LogLevel = "warn"
+	log.Info().Msgf("Loading %d applications", len(Global.Applications))
+
+	for name, app := range Global.Applications {
+		log.Debug().Msgf("Validating %s application config", name)
+
+		// Deprecated: #70: use Config.Log.Level to set up application logging or SecDebugLogLevel to set up Coraza logging
+		if app.LogLevel != "" {
+			log.Warn().Msg("'app.log_level' is skipped. For setting application log level use 'log.level' in the root of configuration.")
```
Comment (sts) [inline config/config.go:77]: add Str("app", name)

File: config/config.go [hunk 8]
```diff
@@ -54,11 +67,20 @@ func InitConfig(file string) error {
...
+		}
+		// Deprecated: #70: use Config.Log.File to set up application logging or SecDebugLog to set up Coraza logging
+		if app.LogFile != "" {
+			log.Warn().Msg("'app.log_level' is skipped. For setting application log file use 'log.file' in the root of configuration.")
```
Comment (sts) [inline config/config.go:81]: add Str("app", name)

File: config/config.go [hunk 9]
```diff
@@ -54,11 +67,20 @@ func InitConfig(file string) error {
...
 
 func validateConfig() error {
-	fmt.Printf("Loading %d applications\n", len(Global.Applications))
-	for _, app := range Global.Applications {
-		if app.LogLevel == "" {
-			app.LogLevel = "warn"
+	log.Info().Msgf("Loading %d applications", len(Global.Applications))
+
+	for name, app := range Global.Applications {
+		log.Debug().Msgf("Validating %s application config", name)
```
Comment (sts) [inline config/config.go:73]: add Str("app", name)

File: docker/haproxy/haproxy.cfg [hunk 10]
```diff
@@ -40,10 +40,14 @@ frontend test_frontend
     http-request silent-drop if { var(txn.coraza.action) -m str drop }
     http-response silent-drop if { var(txn.coraza.action) -m str drop }
 
-    # Deny in case of an error, when processing with the Coraza SPOA
+    # Deny in case of an error, when processing with the Coraza SPOE
```
Comment (sts) [inline docker/haproxy/haproxy.cfg:43]: The SPOE is the Haproxy part which talks to the SPOA. So I think this as correct.

Comment (zc-devs) [reply docker/haproxy/haproxy.cfg:43]: HAProxy talks about [SPOE/SPOP](https://github.com/haproxy/haproxy/blob/49ddd87d41b62797d50d7c59e843d0c0b0236eb5/doc/SPOE.txt#L300-L337) only.

There are `txn.coraza.err_code` and `txn.coraza.err_msg` for SPOA errors now.

**Comment by zc-devs** [comment]
> How about we just log out the first line

Fixed.

> not return in [internal/spoa](https://github.com/corazawaf/coraza-spoa/blob/main/internal/spoa.go#L188) but rather continue

HAProxy doesn't run if there is error in config, doesn't matter if error is in only one backend and others work. So, I would stick with this behavior.

> While in context of an app, the app name should be included

As I sad `Maybe later I will implement loggers by package/'class'`, I would make it in other issue/PR. Both of them introduce new loggers.

> for requests, the request id should also be included in the log line

As I use Coraza's DebugLog with `Str("transaction_id", tx.ID())` and `tx = app.waf.NewTransactionWithID(req.id)`, seems it's done.

**Review by jcchavezs — CHANGES_REQUESTED** [review]
I really don't see the need to have zerolog and logrus in here. Do we need both and also a `log` package? I would prefer to use coraza logger (embedded in the transaction) and std logger for those cases where it is not possible and from there explore the possibility of retrieving the logger from the WAF object or do a recommendation. The reason is this is the only connector where I see we need to introduce new loggers.

**Comment by fionera** [comment]
Rewrite does use zerolog. Closed via #103
